### PR TITLE
[REVIEW] null aware binary op to compare two columns (or) column and a scalar

### LINKED ISSUE #3749: [FEA] null aware equals binary op
**Is your feature request related to a problem? Please describe.**
For spark support we need a SQL op '<=>' which 

> Returns same result as the EQUAL(=) operator for non-null operands, but returns true if both are null, false if one of the them is null.

This is **NOT** the same as the c++ <=> operator.

**Describe the solution you'd like**
I would love to see a new binary op that does what we want.

**Describe alternatives you've considered**
We can probably make this work with a combination of isNull, and copyIf, but that feels overly complicated and would result in a lot more columns being materialized.

### PR DESCRIPTION
**Pull request by sriramch**
- this Closes #3749
  - it returns a non null-able boolean column that returns 1 if both are null/invalid/not_present;
    0 if only one of them is null/invalid/not_present; or the result of comparing the 2 values

### CODE REVIEW

**Comment by sriramch** [comment]
@jrhemstad from what i could tell, the current binops jit mechanics neither distinguishes nor propagates the validity status of the column entries or the scalar to the binary operation function, which is required for this operation. further, the output column needs to be unconditionally non nullable (and not nullable only because one of the operand or its entries are invalid). instead of changing all the existing operations to be validity aware, i special cased this.

alternatively, by adding an enum, i can check if this op is `null_aware_equal` (new enum) and redirect it to a compiled binops implementation that performs this operation without introducing a new api. would that be palatable?

if there are much simpler ways to do this via adding an enum to the binop, please let me know.

File: cpp/src/binaryop/binaryop.cpp [hunk 1]
```diff
@@ -313,8 +322,16 @@ std::unique_ptr<column> binary_operation(column_view const& lhs,
   CUDF_EXPECTS(is_fixed_width(output_type),
                "Invalid/Unsupported output datatype");
 
-  if ((lhs.type().id() == type_id::STRING) &&
-      (rhs.type().id() == type_id::STRING)) {
+  if (op == binary_operator::NULL_AWARE_EQUAL) {
+    // Check for datatype
+    CUDF_EXPECTS(lhs.type() == rhs.type(), "Both columns must be of the same type");
+    CUDF_EXPECTS(lhs.size() == rhs.size(), "Both columns must be of the same size");
```
Comment (devavret) [inline cpp/src/binaryop/binaryop.cpp:328]: Could you move these restrictions to inside `null_aware_equal`.

Comment (sriramch) [reply cpp/src/binaryop/binaryop.cpp]: done.

**Comment by jrhemstad** [comment]
> alternatively, by adding an enum, i can check if this op is `null_aware_equal` (new enum) and redirect it to a compiled binops implementation that performs this operation without introducing a new api. would that be palatable?

That would be preferable, yes.

I'd also like to see different nomenclature. I think the naming of "null aware" is misleading. It makes it seem as though other binary operations are null "unaware". Something like `NULL_EQUALS` or `NULL_VALUE_EQUALS` for the enum would be better.

**Comment by sriramch** [comment]
> Something like `NULL_EQUALS` or `NULL_VALUE_EQUALS` for the enum would be better.

[sc] changed enum to `null_equals`

**Comment by sriramch** [comment]
> later it'll explode compile times if we implement comparisons for different types per @devavret's suggestion:

[sc] a number of binops between different types are invalid anyways (timestamp and integral/floating point types; string and timestamps/integral/floating point types etc.). is there a reason why we support arbitrary combinations? was it just intended to be used between bool/integral/floating types without creating intermediary columns?

> 
> > ...I'd prefer if this new op could also take different typed inputs...
> 
> Is there a reason we can't add this as a new null-aware kernel in the JIT'd code?

[sc] most of these null aware ops have to be special cased, in that the validity of element in the output column may no longer be a simple `and/&` of the validity of the 2 column elements or the column and scalar element. this can be handled in the kernel; but the overall binop operation can't be as generic as it is today. furthermore, handling strings requires special consideration as well. presently, all these string binops pretty much returns only a bool value. this may no longer be the case as some of these binops may want to return one of the operand in its result. hence, i wanted to handle these ops separately.

**Comment by trxcllnt** [comment]
I agree many comparisons don't make sense. My understanding is the binops are JIT compiled to improve compile times since they double-dispatch. IIRC the compiler generates all `num_type_ids^2` possible overload candidates, and only eliminates the ones made invalid through SFINAE during overload resolution.

**Comment by sriramch** [comment]
@trxcllnt i'm not sure if comparing unrelated types are all that useful. `replace_nulls` is somewhat a similar api (that needs to respect the nullity/validity of the columns) that could have been implemented as a binop via the jit mechanics; but it wasn't. perhaps, the generalization or the assumption that the binop jit mechanism makes isn't relevant or applicable (requirement of non nullable output column, enforcement of compatible types etc.). i think this binop falls under a similar category.

**Comment by jrhemstad** [comment]
> i'm not sure if comparing unrelated types are all that useful.

> @trxcllnt i'm not sure if comparing unrelated types are all that useful.

There are many types that are related, such as comparing an `int8` column to an `int64` column, or a `float` to a `double`, or `int32` to `float`,  where we don't have to have to materialize the intermediate results. I'd wager there are more type combinations that are valid than are invalid. 

This is the pattern we use in libcudf for binary ops and it should be followed. 

> replace_nulls is somewhat a similar api (that needs to respect the nullity/validity of the columns) that could have been implemented as a binop via the jit mechanics; but it **wasn't.**

`replaces_nulls` is not a binary op. I don't understand how this relates.

**Comment by sriramch** [comment]
> `replaces_nulls` is not a binary op. I don't understand how this relates.

[sc] what i meant was that `replace_nulls` could have been modeled as a binops where one operand is an input column and another a scalar/column containing the corresponding replacement values. the output value is the non null value between both (or null if both are null). the unimplemented `coalesce` operator could have been the closest. i was only trying to understand why a separate api was chosen for that.

**Comment by trxcllnt** [comment]
> i'm not sure if comparing unrelated types are all that useful

My apologies, I'm not making myself clear. I agree certain comparisons like `int32_t == string` aren't valid. But others, like `int32_t == int64_t`, are. @devavret's suggestion is to support comparisons between similar-ish types like this, and the issue is I'm worried that will negatively impact compilation time.

binops do three nested `type_dispatcher()` calls: one for the out type, another for the lhs type, and a third for the rhs type. You can see this commented out in [an earlier iteration of the compiled binops](https://github.com/rapidsai/cudf/blob/f6ad0a97f2097a48b0070485e824c263edd95b70/cpp/src/binaryop/compiled/binary_ops.cu#L204) to verify how it impacts compile times.

In pseudo-code, it looks like this:
```c++
unique_ptr<column> binary_operation(column_view const &lhs,
                                    column_view const &rhs, binary_operator op,
                                    data_type output_type) {
  // assume these lambdas are actually type-parameterized functors
  return type_dispatcher(output_type, [&]() {
    return type_dispatcher(lhs.type(), [&]() {
      return type_dispatcher(rhs.type() [&]() {
        // do binop, return result column
      });
    });
  });
}
```

The number of overload combinations compiled for each binop is potentially large:
```
num_type_ids = (today) 16
num_rhs_types = num_type_ids
num_lhs_types = num_type_ids
num_out_types = variable based on the binop

// in the range 16^[2,3], i.e. 256-4096 overload combinations
num_overload_combinations = num_out_types * num_lhs_types * num_rhs_types
```

Since this binop only returns bool, we can remove the output type dispatcher call, keeping this on the low-end of only compiling 16^2 overload combinations. Better, but still not great since we're not done adding new types yet.

I'm suggesting adding another function to [`jit/code/kernel.cpp`](https://github.com/rapidsai/cudf/blob/6fde8d3a8e31ff3d5a50bb1e30d35ba2cec72f6d/cpp/src/binaryop/jit/code/kernel.cpp) with a custom validity check instead, like this:
```c++
template <typename TypeOut, typename TypeLhs, typename TypeRhs, typename TypeOpe, typename IsValidFunc>
__global__
void kernel_v_v_check_valid(cudf::size_type size, TypeOut* out_data,
                            TypeLhs* lhs_data, bitmask_type* lhs_mask,
                            TypeRhs* rhs_data, bitmask_type* rhs_mask) {
    int tid = threadIdx.x;
    int blkid = blockIdx.x;
    int blksz = blockDim.x;
    int gridsz = gridDim.x;

    int start = tid + blkid * blksz;
    int step = blksz * gridsz;

    for (cudf::size_type i=start; i<size; i+=step) {
        auto const is_valid = IsValidFunc::template operate<TypeLhs, TypeRhs>(
            lhs_data[i], bit_is_set(lhs_mask, i), rhs_data[i], bit_is_set(rhs_mask, i));
        if (is_valid) {
            out_data[i] = TypeOpe::template operate<TypeOut, TypeLhs, TypeRhs>(lhs_data[i], rhs_data[i]);
        }
    }
}
```

Something like this (or one that also writes the `out_mask`) might also allow us to remove the null mask and-ing/or-ing workarounds we have in Python.

**Comment by sriramch** [comment]
thanks @trxcllnt 

> My apologies, I'm not making myself clear. I agree certain comparisons like `int32_t == string` aren't valid. But others, like `int32_t == int64_t`, are. @devavret's suggestion is to support 

[sc] your earlier comment was very clear and i was asking earlier `was it just intended to be used between bool/integral/floating types without creating intermediary columns?`, which you have now confirmed is required and useful.

> Since this binop only returns bool, we can remove the output type dispatcher call, keeping this on the low-end of only compiling 16^2 overload combinations. Better, but still not great since we're not done adding new types yet.

[sc] there are going to be binops that return something besides bool. for instance, null-aware max/min may return the non-null operand if one is null. the output bitmask and string handling may all require special case handling as well.

> 
> I'm suggesting adding another function to [`jit/code/kernel.cpp`](https://github.com/rapidsai/cudf/blob/6fde8d3a8e31ff3d5a50bb1e30d35ba2cec72f6d/cpp/src/binaryop/jit/code/kernel.cpp) with a custom validity check instead, like this:

[sc] ok - thanks for that. i'd something along these lines this morning and i'll iterate on that.

**Comment by devavret** [comment]
@trxcllnt  Thanks for explaining this in a much better way. Yes, that's what I meant when I argued about supporting differing types. `int16 <=> int8` is still a valid combination. (SQL version of `<=>`)

> I'm suggesting adding another function to [`jit/code/kernel.cpp`](https://github.com/rapidsai/cudf/blob/6fde8d3a8e31ff3d5a50bb1e30d35ba2cec72f6d/cpp/src/binaryop/jit/code/kernel.cpp) with a custom validity check instead, like this:

The reason I didn't request this change is twofold:
1. We are considering making binary ops compiled again. @karthikeyann found a way to do it without exploding compile time and binary size too much.
2. JIT equality doesn't support strings, which I suppose would be required since comparisons are the operators that do support strings in compiled version.

**Comment by trxcllnt** [comment]
> 1. We are considering making binary ops compiled again. @karthikeyann found a way to do it without exploding compile time and binary size too much.
> 2. JIT equality doesn't support strings, which I suppose would be required since comparisons are the operators that do support strings in compiled version.

1. ...!
2. Ugh, I forgot about this. One workaround is to have an early-return if the operator is this new `<=>` one, and dtypes are string + anything else, return a column of all-false :slightly_frowning_face:

**Comment by devavret** [comment]
> 2. Ugh, I forgot about this. One workaround is to have an early-return if the operator is this new `<=>` one, and dtypes are string + anything else, return a column of all-false 🙁

We don't do that now for regular equality. What is Spark's desired behaviour for comparisons of incompatible types. Does that return all null?

**Comment by sriramch** [comment]
> JIT equality doesn't support strings, which I suppose would be required since comparisons are the operators that do support strings in compiled version.

[sc] does the jit mechanics work when one of the operand is a string for *any* binop today? the device (column data) pointer cannot be interpreted like a fixed width type - right? it looks like cudf uses a compiled version *only* when both are strings.

> What is Spark's desired behaviour for comparisons of incompatible types. Does that return all null/false? 

[sc] please see [here](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_equal). if we aren't going to support comparison between incompatible types when one of them is a string, why don't we not support between any? `cudf::fill` could be used even when comparing between integral/floating point types - right?

**Comment by sriramch** [comment]
@trxcllnt @devavret i have incorporated your feedback to now support this binop between unrelated types via the jit mechanics. please take a look. 

a couple of things can be done as a part of a future pr, if needed:

- comparison between strings and non-strings data types (was *unsupported* earlier and still is)
- binop(string, string) returning a string value isn't supported. it isn't required for this pr, but may be required for future

**Comment by sriramch** [comment]
superseded by #5004

**Pull request closed without merging**